influxdb3_write/src/write_buffer/mod.rs

```diff
@@ -1147,6 +1162,110 @@ mod tests {
+                PutPayload::from_bytes(Bytes::from(snapshot_json)),
+            )
+            .await
+            .unwrap();
+
+        // setup the write buffer:
+        let (wbuf, _ctx) = setup(
+            Time::from_timestamp_nanos(0),
+            Arc::clone(&object_store),
+            WalConfig {
+                level_0_duration: Level0Duration::new_1m(),
+                max_write_buffer_size: 100,
+                flush_interval: Duration::from_millis(5),
+                snapshot_size: 1,
+            },
+        )
+        .await;
+
+        // there should be one snapshot already, i.e., the one we created above:
+        verify_snapshot_count(1, &wbuf.persister).await;
+        // there aren't any catalogs yet:
+        verify_catalog_count(0, object_store.clone()).await;
+
+        // do three writes to force a new snapshot
+        wbuf.write_lp(
+            NamespaceName::new("foo").unwrap(),
+            "cpu bar=1",
+            Time::from_timestamp(10, 0).unwrap(),
+            false,
+            Precision::Nanosecond,
+        )
+        .await
+        .unwrap();
+        wbuf.write_lp(
+            NamespaceName::new("foo").unwrap(),
+            "cpu bar=2",
+            Time::from_timestamp(20, 0).unwrap(),
+            false,
+            Precision::Nanosecond,
+        )
+        .await
+        .unwrap();
+        wbuf.write_lp(
+            NamespaceName::new("foo").unwrap(),
+            "cpu bar=3",
+            Time::from_timestamp(30, 0).unwrap(),
+            false,
+            Precision::Nanosecond,
+        )
+        .await
+        .unwrap();
+
+        // Check that there are now 2 snapshots:
+        verify_snapshot_count(2, &wbuf.persister).await;
+        // Check that the next sequence number is used for the new snapshot:
+        assert_eq!(
+            prev_snapshot_seq.next(),
+            wbuf.wal.last_snapshot_sequence_number().await
+        );
+        // There should be a catalog now, since the above writes updated the catalog
+        verify_catalog_count(1, object_store.clone()).await;
+        // Check the catalog sequence number in the latest snapshot is correct:
+        let persisted_snapshot_bytes = object_store
+            .get(&SnapshotInfoFilePath::new(
+                "test_host",
+                prev_snapshot_seq.next(),
+            ))
+            .await
+            .unwrap()
+            .bytes()
+            .await
+            .unwrap();
+        let persisted_snapshot =
+            serde_json::from_slice::<PersistedSnapshot>(&persisted_snapshot_bytes).unwrap();
+        // TODO: it appears that writes which create a new db increment the catalog sequence twice,
+        // hence why the sequence number is 2 here, if we manage to make it so that scenario only
+        // increments the catalog sequence once, then this assertion may fail:
+        assert_eq!(
+            SequenceNumber::new(2),
+            persisted_snapshot.catalog_sequence_number
+        );
```

Comment:
I think this is a bug, the sequence number should be 1 here, but instead it is 2. I believe this is because the write operation is both creating a database _and_ updating the table schema for that database, resulting in calling both [`db_or_create`](https://github.com/influxdata/influxdb/blob/d0720a4fe46458763466c5ebc82b4ad932517f01/influxdb3_catalog/src/catalog.rs#L141) and [`apply_catalog_batch`](https://github.com/influxdata/influxdb/blob/d0720a4fe46458763466c5ebc82b4ad932517f01/influxdb3_catalog/src/catalog.rs#L137), which both increment the sequence number. I can open an issue for this.

Reply:
I think it's fine that the catalog sequence gets incremented twice? It's once for the db create and then other times for the tables. The catalog file path uses the wal sequence number and there are no downstream clients that monitor the catalog. The downstream compactor will have to pull the catalog when it gets updated, but that should be something it can determine from the `PersistedSnapshot` if the catalog sequence number in there is different than the one it knows about.

Reply:
That sounds reasonable to me. I updated the comment above the assertion to just be a note to future selves instead of a `TODO`. I can merge once ✅.